test/common/TestCommonEqEpsilon.hs

```diff
@@ -34,23 +39,136 @@ eqEpsilonRef = unsafePerformIO $ newIORef eqEpsilonDefault
+go_assert_list _ [] [] = assertBool "" True
+go_assert_list _ [] (_:_) = assertFailure "More list elements than expected!"
+go_assert_list _ (_:_) [] = assertFailure "Less list elements than expected!"
+go_assert_list mk (head_exp:tail_exp) (head_act:tail_act) =
+      mk head_exp head_act >> go_assert_list mk tail_exp tail_act
+
+assert_list :: (a -> a -> Assertion) -- ^ The function used to make an assertion on two elements (expected, actual)
+            -> [a]                   -- ^ The expected value
+            -> [a]                   -- ^ The actual value
+            -> Assertion
+assert_list make_assert expected actual =
+  if lenE == lenA then
+    go_assert_list make_assert expected actual
+  else
+    assertFailure $ "List too " ++ (if lenE < lenA then "long" else "short") ++ ": expected " ++ show lenE ++ "elements, but got: " ++ show lenA
+  where
+    lenE :: Int = length expected
+    lenA :: Int = length actual
+
+assert_shape :: forall a b. (HasShape a, Linearizable a b)
+             => (b -> b -> Assertion) -- ^ The function used to make an assertion on two elements (expected, actual)
+             -> a                     -- ^ The expected value
+             -> a                     -- ^ The actual value
+             -> Assertion
+assert_shape make_assert expected actual =
+  if shapeE == shapeA then
+    assert_list make_assert (linearize expected) (linearize actual)
+  else
+    assertFailure $ "Expected shape: " ++ show shapeE ++ ", but got: " ++ show shapeA
+  where
+    shapeE = shapeL expected
+    shapeA = shapeL actual
+
+----------------------------------------------------------------------------
+-- Generic comparisons with explicit error margin
+----------------------------------------------------------------------------
+
 -- | Asserts that the specified actual floating point value is close to the expected value.
 -- The output message will contain the prefix, the expected value, and the
 -- actual value.
 --
 -- If the prefix is the empty string (i.e., @\"\"@), then the prefix is omitted
 -- and only the expected and actual values are output.
-assertClose :: forall a. (Fractional a, Ord a, Show a, HasCallStack)
-            => String -- ^ The message prefix
-            -> a      -- ^ The expected value
-            -> a      -- ^ The actual value
-            -> Assertion
-assertClose preface expected actual = do
-  eqEpsilon <- readIORef eqEpsilonRef
-  assertBool (msg eqEpsilon) (abs (expected-actual) < fromRational eqEpsilon)
+assert_close_eps :: (Num a, Ord a, Show a, HasCallStack)
+                   => String -- ^ The message prefix
+                   -> String -- ^ The message suffix
+                   -> a      -- ^ The error margin
+                   -> a      -- ^ The expected value
+                   -> a      -- ^ The actual value
+                   -> Assertion
+assert_close_eps preface epilogue eqEpsilon expected actual = do
+  assertBool (msg eqEpsilon) (abs (expected-actual) < eqEpsilon)
```

Comment:
Just double-checking: would people expect `<` here or `<=`? With floating point it shouldn't matter much, but it may sometimes.

Reply:
Well it could matter e.g. if `eqEpsilon` is `0`. (But on the other hand, we allow it to be negative, i.e. there is no check.) Well spotted, fixed.